data/virt_autotest/guest_console_monitor.sh

```diff
@@ -0,0 +1,236 @@
+vm_guestnames_inactive_array=$(echo -e ${get_vm_guestnames_inactive})
+get_vm_guestnames=`virsh list  --all | grep "${vm_guestnames_types}" | awk '{print $2}'`
+vm_guestnames_array=$(echo -e ${get_vm_guestnames})
+vmguest=""
+vmguest_failed="0"
+
+#Parse input arguments. -s or -e must have values
+while getopts "seh" OPTION; do
+  case "${OPTION}" in
+    s)
+      START_MONITOR="1"
+      ;;
+    e)
+      END_MONITOR="1"
+      ;;
+    h)
+      help_usage
+      exit 1
+      ;;
+    *)
+      help_usage
+      exit 1
+      ;;
+  esac
+done
+
+#FUNCTION PART
+function get_console() {
+
+local vmguest=$1
+
+expect -c "
+set timeout 60
+
+##hide echo
+log_user 0
+spawn -noecho virsh console ${vmguest}
+
+#wait connection
+sleep 3
+send \"\r\n\r\n\r\n\"
+
+#condition expect
+expect {
+        \"*login:\" {
+                send \"root\r\"
+                exp_continue
+        }
+        -nocase "password:" {
+                send \"novell\r\"
+                exp_continue
+        }
+        \"*:~ #\" {
+                send -- \"ip route get 1\r\"
+        }
+        timeout {
+                send -- \"exit\r\"
+                exp_continue        
+        }                 
+}
+
+## -1 means never timeout
+set timeout -1
+
+expect -re {dev.*\s([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})[^0-9]}
+
+if {![info exists expect_out(1,string)]} {
+        puts \"Match did not happen :(\"
+        exit 1
+}
+set output \$expect_out(buffer)
+expect eof
+
+puts \"\$output\"
+"
+
+}
+
+function open_console() {
+
+    #open guest console
+    for vmguest in ${vm_guestnames_state_running_array[@]};do
+        get_console ${vmguest} >> ${LOG_DIR}/${vmguest}_console_${timestamp} 2>&1 &
+    done 
+}
+
+function close_console() {
+
+    #close guest console
+    echo "kill PIDs $(basename $0)"
+    pid=`ps aux | grep "$(basename $0) -s" | grep -v "grep"| awk '{print $2}'`
+    if [[ -n $pid ]];then
+        kill -9 $pid >> /dev/null 2>&1;sync
+    else
+        echo "PIDs $(basename $0) were killed"
+    fi
+
+    echo "Kill all associated "virsh console" processes"
+    ps aux | grep "virsh console" | grep -v "grep" >> /dev/null 2>&1
+    while [[ $? -eq 0 ]];
+    do
+        #wait for checking all associated "virsh console" processes"
+        sleep 3
+        sub_pid=`ps aux | grep "expect -c" | grep -v "grep"| awk '{print $2}'`
+        if [[ -n $sub_pid ]];then
+            kill -9 $sub_pid >> /dev/null 2>&1;sync
+        else
+            break
+        fi
+    done
+}
+
+function end_monitor() {
+
+    #end of monitor for guest console
+    if [ -f ${MONITOR_LOCKED_FILE} ];then
+        echo "End of Monitor for guest console"	
+        #UNLOCK Monitor for guest console	    
+        > ${MONITOR_LOCKED_FILE} ; rm -rf ${MONITOR_LOCKED_FILE}
+        #Kill Pids from Monitor for guest console
+        close_console
+    fi
+}
+
+function start_monitor() {
```

Comment:
Could you help reboot any running virtual machine before do starting console ? Reboot can help serve virtual machine cleanup functionality which saves use much effort to really do the thing. Just one line "virsh reboot"

Reply:
And after thinking twice, I think you need to start any inactive virtual machine before starting any console also. Also just do this once before starting any console. This can make sure we are in normal situation before go ahead to do real test. Sorry for any inconvenience.

Reply:
Great, this is a good idea. I had thought about this as your mentioned.
But, this script just only handle guest serial console and collect the output from guest serial console without the other operation for vm guest (included virsh reboot), imho. 

Also, we can use this script anytime as our test needed. So, we add this "virsh reboot" to help serve virtual machine cleanup, but we also effect the current running test both on vm guest and host together. 

So,  for now, we'd better to use longer timeout when do login in expect without any operation for vm guest.

Reply:
I think you can help do "virsh reboot" here. It does no harm and can save us much effort to it purposefully. There is no need to do additional guest cleanup. Agree to use longer timeout.

Reply:
OK. No need to add "virsh reboot" anywhere. So also no need to change timeout